File: src/core/Transactions/Handlers/ImportTransactions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using core.Account;
using core.Adapters.CSV;
using core.Adapters.Emails;
using core.Options;
using core.Shared;
using core.Stocks;
using MediatR;

namespace core.Transactions.Handlers
{
    public class Import
    {
        public class Command : RequestWithUserId<CommandResponse>
        {
            public Command(string content)
            {
                Content = content;
            }

            public string Content { get; }
        }

        public class Handler : IRequestHandler<Command, CommandResponse>
        {
            private IEmailService _emailService;
            private IMediator _mediator;
            private ICSVParser _parser;
            private IAccountStorage _storage;

            public Handler(
                IEmailService emailService,
                IMediator mediator,
                ICSVParser parser,
                IAccountStorage storage)
            {
                _emailService = emailService;
                _mediator = mediator;
                _parser = parser;
                _storage = storage;
            }

            public async Task<CommandResponse> Handle(Command request, CancellationToken cancellationToken)
            {
                var parseResponse = _parser.Parse<TransactionRecord>(request.Content);
                if (!parseResponse.IsOk)
                {
                    return CommandResponse.Failed(parseResponse.Error!.Message);
                }

                var user = await _storage.GetUser(request.UserId);
                if (user == null)
                {
                    return CommandResponse.Failed("User not found");
                }

                await SendEmail(user, subject: "Started importing transactions", body: "");

                var errors = new List<string>();
                try
                {
                    foreach (var cmd in GetCommands(parseResponse.Success))
                    {
                        cmd.WithUserId(request.UserId);
                        var response = await _mediator.Send(cmd, cancellationToken);
                        if (response is CommandResponse { Error: not null } r)
                        {
                            errors.Add(r.Error);
                        }
                    }
                }
                catch (Exception ex)
                {
                    await SendEmail(
                        user: user,
                        subject: "Failed to import transactions",
                        body: ex.ToString()
                    );

                    return CommandResponse.Failed($"Entry Failed: {ex}");
                }

                var onFinished = errors.Count switch
                {
                    0 => AfterSuccess(user),
                    _ => AfterFailure(errors, user)
                };

                return await onFinished;
            }

            private async Task<CommandResponse> AfterFailure(List<string> errors, User user)
            {
                var body = string.Join(",", errors);

                await _emailService.Send(
                    recipient: new Recipient(email: user.State.Email, name: user.State.Name),
                    sender: Sender.NoReply,
                    subject: "Failed to import transactions",
                    body: body
                );

                return CommandResponse.Failed(body);
            }

            private async Task<CommandResponse> AfterSuccess(User user)
            {
                await _emailService.Send(
                    recipient: new Recipient(email: user.State.Email, name: user.State.Name),
                    sender: Sender.NoReply,
                    subject: "Finished importing transactions",
                    body: ""
                );

                return CommandResponse.Success();
            }

            private Task SendEmail(User user, string subject, string body) =>
                _emailService.Send(
                    recipient: new Recipient(email: user.State.Email, name: user.State.Name),
                    sender: Sender.NoReply,
                    subject: subject,
                    body: body
                );

            private IEnumerable<RequestWithUserIdBase> GetCommands(IEnumerable<TransactionRecord> records)
            {
                // transaction records come in sorted by descending date
                foreach(var record in records.Reverse().Where(t => t.Qualify))
                {
                    if (record.IsStock())
                    {
                        yield return CreateStockTransaction(record);
                    }
                    else if (record.IsOption())
                    {
                        yield return CreateOptionTransaction(record);
                    }
                }
            }

            private StockTransaction CreateStockTransaction(TransactionRecord record) => record.IsBuy() switch
            {
                true => new core.Stocks.Buy.Command { Ticker = record.Symbol, Date = record.Date, Notes = record.Description, NumberOfShares = record.Quantity.Value, Price = record.Price.Value },
                false => new core.Stocks.Sell.Command { Ticker = record.Symbol, Date = record.Date, Notes = record.Description, NumberOfShares = record.Quantity.Value, Price = record.Price.Value }
            };

            private OptionTransaction CreateOptionTransaction(TransactionRecord record) => record.IsBuy() switch
            {
                true => new core.Options.Buy.Command { Ticker = record.GetTickerFromOptionDescription(), Filled = record.Date, Notes = record.Description, NumberOfContracts = Decimal.ToInt32(record.Quantity.Value), Premium = record.Price.Value * 100, StrikePrice = record.StrikePrice(), ExpirationDate = record.ExpirationDate(), OptionType = record.OptionType() },
                false => new core.Options.Sell.Command { Ticker = record.GetTickerFromOptionDescription(), Filled = record.Date, Notes = record.Description, NumberOfContracts = Decimal.ToInt32(record.Quantity.Value), Premium = record.Price.Value * 100, StrikePrice = record.StrikePrice(), ExpirationDate = record.ExpirationDate(), OptionType = record.OptionType() }
            };

            private class TransactionRecord
            {
                public DateTimeOffset Date { get; set; }
                public string TransactionId { get; set; }
                public string Description { get; set; }
                public decimal? Quantity { get; set; }
                public string Symbol { get; set; }
                public decimal? Price { get; set; }
                internal bool Qualify => IsBuy() || IsSell();
                internal bool IsBuy() => Description.StartsWith("Bought ");
                internal bool IsSell() => Description.StartsWith("Sold ");
                internal bool IsOption() => _optionContractRegex.IsMatch(Description);
                internal bool IsStock() => !IsOption();

                // LOB Nov 19 2021 60.0 Call
                private static readonly Regex _optionContractRegex = new Regex(@"(\w+) (\w+ \d+ \d+) (\d{1,6}.\d{1,4}) (Put|Call)");
                internal decimal StrikePrice()
                {
                    var match = _optionContractRegex.Match(Description);

                    return Decimal.Parse(match.Groups[3].Value);
                }

                internal DateTimeOffset? ExpirationDate()
                {
                    var match = _optionContractRegex.Match(Description);

                    return DateTimeOffset.ParseExact(match.Groups[2].Value, "MMM dd yyyy", null);
                }

                internal string OptionType()
                {
                    var match = _optionContractRegex.Match(Description);

                    return match.Groups[4].Value;
                }

                internal string GetTickerFromOptionDescription()
                {
                    var match = _optionContractRegex.Match(Description);

                    return match.Groups[1].Value;
                }
            }
        }
    }
}

File: src/core.fs/ImportTransactions.fs
namespace core.fs

open System
open System.Text.RegularExpressions
open core.Shared
open core.fs.Accounts
open core.fs.Adapters.Brokerage
open core.fs.Adapters.CSV
open core.fs.Adapters.Email
open core.fs.Adapters.Storage
open core.fs.Options
open core.fs.Stocks

module ImportTransactions =
    
    type Command =
        {
            Content:string
            UserId:UserId
        }
        
    type internal TransactionRecord =
        {
            Date:DateTimeOffset
            TransactionId:string
            Description:string
            Quantity:Nullable<decimal>
            Symbol:string
            Price:Nullable<decimal>
        }
        
        static member OptionRegex = Regex(@"(\w+) (\w+ \d+ \d+) (\d{1,6}.\d{1,4}) (Put|Call)")
        
        member this.IsBuy() =
            this.Description.StartsWith("Bought ")
            
        member this.IsSell() =
            this.Description.StartsWith("Sold ")
            
        member this.IsOption() =
            TransactionRecord.OptionRegex.IsMatch(this.Description)
        
        member this.IsStock() =
            this.IsOption() |> not
        
        member this.Qualify() =
            this.IsBuy() || this.IsSell()
            
        member this.StrikePrice() =
            let m = TransactionRecord.OptionRegex.Match(this.Description)
            Decimal.Parse(m.Groups[3].Value)
            
        member this.ExpirationDate() =
            let m = TransactionRecord.OptionRegex.Match(this.Description)
            DateTimeOffset.ParseExact(m.Groups[2].Value, "MMM dd yyyy", null)
            
        member this.OptionType() =
            let m = TransactionRecord.OptionRegex.Match(this.Description)
            OptionType.FromString m.Groups[4].Value
            
        member this.GetTickerFromOptionDescription() =
            let m = TransactionRecord.OptionRegex.Match(this.Description)
            m.Groups[1].Value
            
    type Handler(
        accounts:IAccountStorage,
        emailService:IEmailService,
        csvParser:ICSVParser,
        optionsImport:core.fs.Options.OptionsHandler,
        stocksImport:core.fs.Portfolio.StockPositionHandler) =
        
        interface IApplicationService
        
        member _.Handle(cmd:Command) = task {
            
            let sendEmail (user:User) subject body =
                emailService.Send (Recipient(email=user.State.Email, name=user.State.Name)) Sender.NoReply subject body null
            
            let createOptionTransaction r =
                {
                    Filled = Nullable<DateTimeOffset> r.Date
                    Notes = r.Description
                    NumberOfContracts = int r.Quantity.Value
                    Premium = Nullable<decimal> (r.Price.Value * 100m)
                    StrikePrice = Nullable<decimal>(r.StrikePrice())
                    ExpirationDate = Nullable<DateTimeOffset>(r.ExpirationDate())
                    OptionType = r.OptionType()
                    Ticker = Ticker(r.GetTickerFromOptionDescription())
                }
                
            let createStockTransaction r : core.fs.Portfolio.StockTransaction =
                {
                    PositionId = StockPositionId.create()
                    Date = Some r.Date
                    NumberOfShares = r.Quantity.Value
                    Price = r.Price.Value
                    StopPrice = None
                    BrokerageOrderId = None
                }
            
            let! user = accounts.GetUser(cmd.UserId)
            match user with
            | None -> return "User not found" |> ServiceError |> Error
            | Some user ->
                
                let records = cmd.Content |> csvParser.Parse<TransactionRecord>
                match records with
                | Error err -> return Error err
                | Ok records ->
                    let! emailResponse = sendEmail user "Started importing transactions" "Email will follow when complete."

                    match emailResponse with
                    | Error err -> return ServiceError err |> Error
                    | Ok () ->
                    
                    let! processed =
                        records
                        |> Seq.rev
                        |> Seq.filter(_.Qualify())
                        |> Seq.map(fun r -> async {
                            match r.IsOption(),r.IsStock(),r.IsBuy(),r.IsSell() with
                            | true, _, true, _ -> // buy option
                                // let data = r |> createOptionTransaction
                                // let cmd = BuyOrSellCommand.Buy(data, cmd.UserId)
                                // let! result = optionsImport.Handle(cmd) |> Async.AwaitTask
                                // TODO: implement
                                return Ok ()
                            | true, _, _, true -> // sell option
                                // let data = r |> createOptionTransaction
                                // let cmd = BuyOrSellCommand.Sell(data, cmd.UserId)
                                // let! result = optionsImport.Handle(cmd) |> Async.AwaitTask
                                // return result |> Result.bind (fun _ -> Ok ())
                                // TODO: implement
                                return Ok()
                            | _, true, true, _ -> // buy stock
                                let data = r |> createStockTransaction 
                                let cmd = core.fs.Portfolio.Buy(data, cmd.UserId)
                                return! stocksImport.Handle(cmd) |> Async.AwaitTask
                            | _, true, _, true -> // sell stock
                                let data = r |> createStockTransaction
                                let cmd = core.fs.Portfolio.Sell(data, cmd.UserId)
                                return! stocksImport.Handle(cmd) |> Async.AwaitTask
                            | _ -> return Ok ()
                        })
                        |> Async.Sequential
                        |> Async.StartAsTask
                    
                    let okOrError = processed |> ResponseUtils.toOkOrConcatErrors
                    
                    match okOrError with
                    | Ok _ -> 
                        let! emailResult = sendEmail user "Finished importing transactions" "Finished importing your transactions."
                        match emailResult with
                        | Error err -> return ServiceError err |> Error
                        | Ok () -> return Ok ()
                    | Error err ->
                        return $"Failed to import transactions: {err.Message}" |> ServiceError |> Error
        }
